Bump Gateway API Inference Extension (GIE) to v1.2.0

sbekkerm opened the pull request:
## What this PR does / why we need it:

This PR upgrades Gateway API Inference Extension (GIE) to v1.2.0 and migrates the LLMInferenceService API from `serving.kserve.io/v1alpha1` to `serving.kserve.io/v1alpha2`.

### Background

The Gateway API Inference Extension (GIE) project released v1.2.0, which graduates their API from `api/v1alpha2` to `api/v1`. This is a breaking change in the upstream dependency that affects KServe's LLMInferenceService, which embeds GIE types like `InferencePoolSpec`.

Key upstream changes in GIE v1.2.0:
- API package moved from `api/v1alpha2` → `api/v1`
- API group graduated from `inference.networking.x-k8s.io` → `inference.networking.k8s.io`
- Resource renamed from `InferenceModel` → `InferenceObjective`

### Changes in this PR

**GIE Dependency Upgrade**
- Updated `gateway-api-inference-extension` from v0.3.0 to v1.2.0
- Updated import paths from `api/v1alpha2` to `api/v1`

**LLMInferenceService API Version Bump (v1alpha1 → v1alpha2)**
- The KServe API version bump reflects the breaking upstream GIE dependency change
- Migrated all controllers, clients, informers, and listers to v1alpha2
- Regenerated CRDs and OpenAPI schemas
- Added conversion webhooks for v1alpha1 ↔ v1alpha2 backward compatibility

**API Group Alignment**
- Updated HTTPRoute backendRef to use graduated API group `inference.networking.k8s.io`
- Updated all references from `inference.networking.x-k8s.io` to `inference.networking.k8s.io`

**RBAC Updates**
- Added permissions for both API groups (`k8s.io` and `x-k8s.io`) for backward compatibility
- Updated resource name from `inferencemodels` to `inferenceobjectives`

**InferencePool Readiness Check Improvement**
- Modified `IsInferencePoolReady()` to handle cases where no Gateway controller populates `status.Parents`
- The pool is now considered ready if it exists with a valid selector and target ports
- This fixes an issue where InferencePools were incorrectly marked as "not ready" when using Envoy Gateway with AI Gateway extension

**Workload Pod Label Fix**
- Added missing `app` label to workload pods for InferencePool selector matching
- The InferencePool selector expects pods with label `app: <name>-predictor`, but workload pods were previously missing this label

**llm-d Image Updates**
- Updated llm-d-cuda to v0.4.0
- Updated llm-d-inference-scheduler to v0.4.0
- Updated llm-d-routing-sidecar to v0.4.0

---

## Which issue(s) this PR fixes:

Fixes #4859
Fixes #4823

---

## Type of changes

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### ⚠️ Breaking Change Notice

This PR changes the LLMInferenceService API version from `v1alpha1` to `v1alpha2`.

**Migration required:**

Before:
```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LLMInferenceService
```

After:
```yaml
apiVersion: serving.kserve.io/v1alpha2
kind: LLMInferenceService
```

**Backward Compatibility:** Conversion webhooks are included to automatically convert between v1alpha1 and v1alpha2. Existing v1alpha1 resources will continue to work, but users are encouraged to migrate to v1alpha2.

---

## Feature/Issue validation/testing:

- [x] Deploy LLMInferenceService with the updated API version (v1alpha2)
- [x] Verify HTTPRoute is accepted by Envoy Gateway
- [x] Integration tests updated and passing
- [x] `make precommit` passes

---

## Special notes for your reviewer:

**Why v1alpha2?** The LLMInferenceService API embeds types from GIE. When GIE graduated their API from `api/v1alpha2` to `api/v1`, it became a breaking change for KServe. The version bump to v1alpha2 signals this incompatibility to users.

**API Group Support:** Both `inference.networking.k8s.io` (graduated) and `inference.networking.x-k8s.io` (experimental) are supported in RBAC for environments that may still have the old CRDs.

**Conversion Webhooks:** This PR includes conversion webhooks that enable backward compatibility between v1alpha1 and v1alpha2. The v1alpha2 API is the hub (storage) version.

**InferencePool Readiness:** The `IsInferencePoolReady` fix addresses a real-world issue where Envoy Gateway + AI Gateway extension doesn't populate `status.Parents` on InferencePools, but the pool is still functional when the HTTPRoute resolves successfully.

**Pod Label Fix:** The addition of the `app` label to workload pods fixes InferencePool selector matching, ensuring traffic can be properly routed to predictor pods.

**Large diff warning:** The diff is large due to:
- Package rename from v1alpha1 to v1alpha2
- Regenerated client code, CRDs, and install manifests
- Updated GIE CRDs embedded in install scripts
- Conversion webhook code for backward compatibility

---

## Checklist:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

---

## Release note

```release-note
Upgrade Gateway API Inference Extension to v1.2.0 and migrate LLMInferenceService API from v1alpha1 to v1alpha2. This is a breaking change - update your LLMInferenceService manifests to use `apiVersion: serving.kserve.io/v1alpha2`. Conversion webhooks provide backward compatibility for existing v1alpha1 resources. Also includes llm-d images update to v0.4.0 and fixes for InferencePool readiness checks and pod label selector matching.
```

sbekkerm committed:
Bump gateway-inference-extension to v1alpha2 API

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Fix InferencePool API group mismatch and improve readiness check

- Update HTTPRoute backendRef to use inference.networking.k8s.io instead of
  inference.networking.x-k8s.io to match the graduated GIE API
- Update RBAC to use inferenceobjectives instead of inferencemodels
- Improve IsInferencePoolReady to handle cases where no Gateway controller
  populates status.Parents - consider pool ready if it has valid selector
  and target ports
- Update test fixtures to use correct API group

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Add missing app label to workload pods for InferencePool selector

The InferencePool selector expects pods with label 'app: <name>-predictor',
but the workload pods were missing this label. This caused the InferencePool
to not find any pods to route traffic to.

Added the 'app' label to:
- singleNodeLabels() for single-node deployments
- leaderLabels in expectedMainMultiNodeLWS() for multi-node leaders
- workerLabels in expectedMainMultiNodeLWS() when no leader template exists
- leaderLabels in expectedPrefillMultiNodeLWS() for prefill leaders
- workerLabels in expectedPrefillMultiNodeLWS() when no leader template exists

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sivanantha321 commented on pkg/apis/serving/v1alpha2/llm_inference_service_types.go:
```diff
@@ -219,8 +214,8 @@ type IngressSpec struct {
 //
-// Step 1: Gateway (Envoy) &lt;-- ExtProc --&gt; EPP (select the optimal replica to handle the request)
-// Step 2: Gateway (Envoy) &lt;-- forward request --&gt; Inference Pod X
+// Step 1: Gateway (Envoy) <-- ExtProc --> EPP (select the optimal replica to handle the request)
```
the comment were using html representation to avoid errors when auto generating api docs for the website.

sbekkerm replied:
Thanks, addressed this in the latest commits

sbekkerm committed:
Fix html entity encoding in SchedulerSpec comments

Use html entities (&lt; and &gt;) for arrow characters in comments
to avoid errors when auto-generating API docs for the website.

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Update go.mod dependencies

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Add support for backward compatibility

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Bump llm-d images to v0.4.0

- Update llm-d-dev to llm-d-cuda v0.4.0
- Update llm-d-inference-scheduler to v0.4.0
- Update llm-d-routing-sidecar to v0.4.0

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

pierDipi commented on pkg/apis/serving/v1alpha1/llm_inference_service_types.go:
```diff
@@ -235,10 +248,6 @@ type SchedulerSpec struct {
 // InferencePoolSpec defines the configuration for an InferencePool.
 // 'Spec' and 'Ref' are mutually exclusive.
 type InferencePoolSpec struct {
-	// Spec defines an inline InferencePool specification.
-	// +optional
-	Spec *igwapi.InferencePoolSpec `json:"spec,omitempty"`
```
This is a breaking API change, we need to keep v1alpha1 with the same API structure as before in order to upgrade safely

pierDipi replied:
https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/release-1.2/apix/v1alpha2/inferencepool_types.go#L51-L52

pierDipi commented on pkg/controller/v1alpha2/llmisvc/scheduler.go:
```diff
@@ -35,23 +35,23 @@ import (
-	}
+	// if err := r.reconcileSchedulerInferenceModel(ctx, llmSvc); err != nil {
+	// 	return err
+	// }
```
This can be removed

pierDipi replied:
Same for the related commented functions down

natifridman replied:
removed

pierDipi commented on pkg/controller/v1alpha2/llmisvc/validation/config_validator.go:
```diff
@@ -47,14 +47,14 @@ var _ webhook.CustomValidator = &LLMInferenceServiceConfigValidator{}
 
 func (l *LLMInferenceServiceConfigValidator) SetupWithManager(mgr ctrl.Manager) error {
```
we need to have/keep validators for each version, maybe a good idea to move the validators to apis/serving/v1alpha1/validation and apis/serving/v1alpha2/validation packages (separate packages from API types but versioned) ?

pierDipi commented on pkg/controller/v1alpha2/llmisvc/workload_multi_node.go:
```diff
@@ -239,18 +241,20 @@ func (r *LLMISVCReconciler) expectedMainMultiNodeLWS(ctx context.Context, llmSvc
 		workerLabels["llm-d.ai/role"] = "prefill"
 	}
 	leaderLabels := map[string]string{
+		"app":                         kmeta.ChildName(llmSvc.GetName(), "-predictor"),
```
These seems unrelated changes, maybe we can have a separate PR/follow up for this? 
Generally for LLMs "predictor" is a term we probably want to avoid

pierDipi commented on charts/kserve-llmisvc-resources/templates/webhookconfiguration.yaml:
```diff
@@ -10,7 +10,7 @@ webhooks:
         name: llmisvc-webhook-server-service
         namespace: {{ .Release.Namespace }}
-        path: /validate-serving-kserve-io-v1alpha1-llminferenceservice
+        path: /validate-serving-kserve-io-v1alpha2-llminferenceservice
```
see comment on validating both versions

pierDipi commented on charts/kserve-llmisvc-resources/templates/webhookconfiguration.yaml:
```diff
@@ -38,7 +39,7 @@ webhooks:
         name: llmisvc-webhook-server-service
         namespace: {{ .Release.Namespace }}
-        path: /validate-serving-kserve-io-v1alpha1-llminferenceserviceconfig
+        path: /validate-serving-kserve-io-v1alpha2-llminferenceserviceconfig
```
see comment on validating both versions

pierDipi commented on pkg/apis/serving/v1alpha1/llm_inference_service_conversion.go:
```diff
@@ -0,0 +1,335 @@
+		if src.Scheduler.Pool != nil {
+			dst.Scheduler.Pool = &InferencePoolSpec{
+				Ref: src.Scheduler.Pool.Ref,
+				// Note: v1alpha2 InferencePoolSpec.Spec is not converted (lossy)
```
See comment on API breaking change, we need lossless conversion

pierDipi replied:
There is a conversion method here we could try to re-use https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/release-1.2/apix/v1alpha2/inferencepool_conversion.go#L29

pierDipi reviewed (commented):
Thanks! I did a first pass and left a few comments

For testing, we might want to also augment the testing matrix to test both versions end to end https://github.com/kserve/kserve/blob/f15792d42f31c499cbedc089b9b9b7f8fbcdf6e5/test/e2e/llmisvc/test_llm_inference_service.py#L63-L83

pierDipi commented:
/cc @bartoszmajsak

sbekkerm committed:
Restore v1alpha1 API structure and update conversion

- Restore v1alpha1 LLMInferenceService API to match original structure
- Update v1alpha1 to v1alpha2 conversion to use built-in igwapi conversion
- Use gateway-api-inference-extension InferencePool.ConvertTo/ConvertFrom
  for InferencePoolSpec conversion between apix/v1alpha2 and api/v1
- Regenerate deepcopy and CRD manifests

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Revert unrelated changes from v1alpha2 API bump

- Remove 'app' label with '-predictor' suffix from workload labels.
- Remove commented out InferenceModel/InferenceObjective code from scheduler

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

bartoszmajsak commented on pkg/apis/serving/v1alpha1/llm_inference_service_types.go:
```diff
@@ -130,7 +130,7 @@ type LLMModelSpec struct {
 	// This is used by the Inference Gateway scheduler.
 	// +optional
-	Criticality *igwapi.Criticality `json:"criticality,omitempty"`
+	Criticality *string `json:"criticality,omitempty"`
```
Probably we want to keep the enum validation rules - reusing existing type sounds like a good idea here:

https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/release-0.5/api/v1alpha2/inferencemodel_types.go#L132

natifridman replied:
It was dropped in [v1](https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/release-1.2/api/v1), do we still want to use the enum?

bartoszmajsak replied:
We do have validating webhook for v1alpha1 so probably won't harm to keep the contract. Maybe we can just copy this type internally.

natifridman replied:
Done. I just copied it here

bartoszmajsak commented on pkg/apis/serving/v1alpha1/llm_inference_service_conversion.go:
```diff
@@ -0,0 +1,375 @@
+	dst := v1alpha2.LLMModelSpec{
+		URI:  src.URI,
+		Name: src.Name,
+		// Note: Criticality field is dropped in v1alpha2
```
We are loosing this value when serving `v1alpha1`. One option we discussed in the past was to carry it using annotation so it will not get lost.

natifridman replied:
I pushed a proposal for using annotation in this [commit](https://github.com/kserve/kserve/pull/4886/commits/c62e0be1826ac5564c4a191db46e7c6127b1f0a7)

bartoszmajsak reviewed (commented):
Few minor things to follow-up.

It would be great to expand tests as [@pierDipi suggested](https://github.com/kserve/kserve/pull/4886#pullrequestreview-3570985094).

natifridman committed:
Remove commented code

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Add LLMInferenceService conversion tests and enhance conversion logic

- Introduced unit tests for LLMInferenceService and LLMInferenceServiceConfig to ensure criticality values are preserved during conversions between v1alpha1 and v1alpha2.
- Implemented logic to save and restore criticality values in annotations to prevent data loss during conversion.
- Updated conversion methods to handle criticality and LoRA adapter criticalities appropriately.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Implement compile-time interface compliance checks for LLMInferenceService

- Added compile-time checks to ensure LLMInferenceService and LLMInferenceServiceConfig implement the Convertible and Hub interfaces, respectively.
- This enhancement improves type safety and ensures proper conversion functionality between v1alpha1 and v1alpha2.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Refactor criticality handling in LLMInferenceService and related tests

- Updated criticality handling in LLMInferenceService and LLMInferenceServiceConfig to use a defined Criticality type instead of string.
- Modified conversion logic to ensure criticality values are correctly converted to and from annotations.
- Enhanced unit tests to reflect changes in criticality type and verify proper preservation during conversions.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Add multi-version validation for LLMInferenceService

Move validators from controller package to API packages following
kubebuilder multi-version best practices and existing kserve patterns.

Each API version now has its own validator that works directly with its own types, with separate webhook paths:
- v1alpha1: /validate-serving-kserve-io-v1alpha1-llminferenceservice
- v1alpha2: /validate-serving-kserve-io-v1alpha2-llminferenceservice

Changes:
- Create v1alpha1 validators in pkg/apis/serving/v1alpha1/
- Create v1alpha2 validators in pkg/apis/serving/v1alpha2/
- Add LLMInferenceServiceGVK to v1alpha1 package
- Add ParallelismSpec helper methods to v1alpha1
- Update main.go to register both version validators
- Update webhook manifests for versioned webhooks
- Remove duplicate validation packages from controller directory

Signed-off-by: Nati Fridman <nafridma@redhat.com>

bartoszmajsak commented on cmd/llmisvc/main.go:
```diff
@@ -167,17 +167,29 @@ func main() {
+		os.Exit(1)
+	}
+
+	v1alpha1ConfigValidator := &v1alpha1.LLMInferenceServiceConfigValidator{}
```
Aren't we missing the validation logic from `pkg/controller/v1alpha1/llmisvc/validation/`? We should bring it back and update it for the new version. Up to you if you want to pass it as a closure or extract it as its own func. Let's also restore the tests to improve our safety net.

The duplication in `pkg/controller/v1alpha1/llmisvc/webhook/` should stay deleted. Not sure how that happened in the first place :)

natifridman replied:
Thanks, I brought it back

bartoszmajsak reviewed (changes requested):
Thanks for the hard work, @natifridman and @sbekkerm. We're almost there!

It would also be good to re-run `kube:builder` generation to make sure our manifests reflect latest changes.

bartoszmajsak commented:
@yuzisun @terrytangyuan, as it's quite a significant piece of work, I would very much appreciate the review.

I also wonder if we can cut the release with those changes? it would help with the adoption, as we are behind with GIE/llm-d. If that's an option do we need release notes and docs update?

terrytangyuan commented:
> I also wonder if we can cut the release with those changes? it would help with the adoption, as we are behind with GIE/llm-d. If that's an option do we need release notes and docs update?

Yeah we can probably cut a release and will need some docs update. cc @yuzisun

cabrinha commented:
I've deployed this with the help of @andresllh and it seems to be working well.

Couple things I had to change in the Kustomize manifests:

- ValidatingWebhookConfigurations should only contain 1 item:
  - The old `name: llminferenceservice.kserve-webhook-server.v1alpha1.validator` should be removed entirely
  - The existing `ca-injection-patches` need to be updated to reflect the new name: https://github.com/kserve/kserve/blob/master/config/llmisvc/llmisvc_validatingwebhook_cainjection_patch.yaml#L8 (`llminferenceserviceconfig.kserve-webhook-server.v1alpha2.validator`, etc.)

terrytangyuan commented:
Thanks! @cabrinha 

@sbekkerm Could you address the comments and incorporate the additional changes mentioned by @cabrinha above?

natifridman committed:
Restore LLMInferenceServiceConfig validation logic and tests

Wire up ConfigValidationFunc in main.go to perform template validation
using LoadConfig() and ReplaceVariables() for both v1alpha1 and v1alpha2
config validators. This restores the validation logic that was previously
in pkg/controller/v1alpha1/llmisvc/validation/.

Changes:
- Add ConfigValidationFunc and WellKnownConfigChecker to config validators
- Restore warnings when modifying/deleting well-known configs
- Add validation integration tests for LLMInferenceService webhooks
- Add config validation integration tests for template validation
- Fix v1alpha1 -> v1alpha2 references in controller_int_test.go

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Remove v1alpha1 webhook entries and fix CA injection patch names

Since the LLMInferenceService API has been migrated to v1alpha2, the
v1alpha1 validating webhooks are no longer needed. This removes the
duplicate webhook entries and updates the CA injection patches to
reference the correct v1alpha2 validator names.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman commented:
Thanks @bartoszmajsak and @cabrinha for the comments!
Please let me know if I made the changes as expected

natifridman commented:
@terrytangyuan do we need to make the docs update as part of this PR?
If so, please let me know where and I'll try to make the changes

yuzisun commented:
> @terrytangyuan do we need to make the docs update as part of this PR? If so, please let me know where and I'll try to make the changes

Can you help update the dependency here https://kserve.github.io/website/docs/admin-guide/kubernetes-deployment-llmisvc?

yuzisun commented:
> > I also wonder if we can cut the release with those changes? it would help with the adoption, as we are behind with GIE/llm-d. If that's an option do we need release notes and docs update?
> 
> Yeah we can probably cut a release and will need some docs update. cc @yuzisun

Let's do a Christmas release for 0.17.0 :santa:

terrytangyuan commented:
Great idea

natifridman commented:
> Can you help update the dependency here https://kserve.github.io/website/docs/admin-guide/kubernetes-deployment-llmisvc?

@yuzisun sure. I created [this](https://github.com/kserve/website/pull/572) PR in the website repo

gavrissh commented:
> Let's do a Christmas release for 0.17.0 🎅

Agree, it is a good idea to get this in a release.

sivanantha321 commented on install/v0.16.0/kserve-cluster-resources.yaml:
```diff
@@ -1592,7 +1592,7 @@ spec:
           - "9002"
           - --grpcHealthPort
           - "9003"
-          image: ghcr.io/llm-d/llm-d-inference-scheduler:v0.2.0
```
We don't need to modify the manifests under install directory. These are auto generated when creating a new release.

natifridman replied:
Thanks, reverted this change

natifridman committed:
Revert changes to auto-generated install manifests

The install/ directory files are auto-generated during release by
hack/generate-install.sh. Manual modifications are unnecessary as
the source files in config/ are already updated.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

sivanantha321 commented on config/llmisvcconfig/config-llm-scheduler.yaml:
```diff
@@ -48,22 +56,22 @@ spec:
               timeoutSeconds: 1
             args:
-              - --poolName
+              - --pool-name
```
This renaming should also be done in helm chart as well ?

natifridman replied:
Done

natifridman committed:
Use kebab-case for scheduler args in helm chart

Update llm-d-inference-scheduler CLI flags to use kebab-case naming
convention (e.g., --pool-name instead of --poolName) in the helm chart
to match config/llmisvcconfig/config-llm-scheduler.yaml.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

spolti commented:
Guess we also need to update this file https://github.com/kserve/kserve/blob/master/kserve-deps.env

natifridman commented on kserve-deps.env:
```diff
@@ -35,6 +35,6 @@ OPENTELEMETRY_OPERATOR_VERSION=0.113.0
 LWS_VERSION=v0.7.0
-GATEWAY_API_VERSION=v1.2.1
-GIE_VERSION=v0.3.0
+GATEWAY_API_VERSION=v1.3.1-0.20251106052652-079e4774d76b
+GIE_VERSION=v1.2.0
```
@spolti I think it was updated here. Is there anything else needs to be updated?

spolti replied:
No, the file filter in the PR didn't show it for me, thanks for confirming :)

bartoszmajsak reviewed (commented):
I wonder if `v1alpha1` validation is working or is it bypassed? The `config/webhook/llmisvc/manifests.yaml` only lists `apiVersions: v1alpha2`. Either we need CRD conversion webhook enabled (so `v1alpha1` gets converted before validation), or separate webhook entries for `v1alpha1`. In the `main.go` we actually wire both webhooks so I am confused and maybe missing something here.

Did we test this path, e.g., by trying to create an invalid `v1alpha1` resource (e.g. wrong enum values as a simple unhappy path)?

natifridman committed:
Add v1alpha1 webhook support to llmisvc helm chart

  - Add v1alpha1 webhook entries for LLMInferenceService and
    LLMInferenceServiceConfig in kserve-llmisvc-resources chart
  - Remove conflicting llminferenceservice webhook entries from
    kserve-resources chart that routed to the wrong service

  Previously, the llmisvc helm chart only included v1alpha2 webhooks,
  causing v1alpha1 validation requests to fail. The kserve-resources
  chart also defined these webhooks but routed them to
  kserve-webhook-server-service instead of llmisvc-webhook-server-service,
  causing conflicts when both charts were installed.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman commented:
I tested it after bringing back the webhook
```
kubectl apply -n nati-kserve-demo -f - <<EOF
apiVersion: serving.kserve.io/v1alpha1
kind: LLMInferenceService
metadata:
  name: invalid-parallelism
spec:
  model:
    uri: hf://facebook/opt-125m
    name: facebook/opt-125m
  replicas: 1
  parallelism:
    pipeline: 4
    data: 4
    dataLocal: 2
  router:
    route: {}
    gateway: {}
EOF
The LLMInferenceService "invalid-parallelism" is invalid: spec.parallelism: Invalid value: {"pipeline":4,"data":4,"dataLocal":2}: cannot set both pipeline parallelism and data parallelism (data or dataLocal) simultaneously
```

natifridman committed:
Add v1alpha1 webhook support to llmisvc kustomize

- Add v1alpha1 webhook entries in kustomize config/webhook/llmisvc/manifests.yaml

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Refactor llmisvc config validation to reduce code duplication

Extract common validation logic into validateLLMISVCConfig helper function
and replace duplicate anonymous WellKnownConfigChecker functions with a
shared named function.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

bartoszmajsak commented:
We could probably simplify the validation wiring in the main.go further but overall LGTM. 

I don't see e2e tests, but maybe we can do it as a follow-up?

WDYT @terrytangyuan @yuzisun ?

terrytangyuan commented:
Sounds good

terrytangyuan commented on go.mod:
```diff
@@ -20,67 +20,80 @@ require (
-	sigs.k8s.io/gateway-api v1.2.1
-	sigs.k8s.io/gateway-api-inference-extension v0.3.0
+	sigs.k8s.io/controller-runtime v0.22.4
+	sigs.k8s.io/gateway-api v1.3.1-0.20251106052652-079e4774d76b
```
CI is failing because this tag is missing https://github.com/kubernetes-sigs/gateway-api/tags

terrytangyuan replied:
> [INFO] Installing Gateway API CRDs v1.3.1-0.20251106052652-079e4774d76b... error: unable to read URL "https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.3.1-0.20251106052652-079e4774d76b/standard-install.yaml", server reported 404 Not Found, status code=404

natifridman replied:
@terrytangyuan This version is required for GIE https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/52ed3abfd4aa2374cc0f4ab5abe94675eb35311c/go.mod#L39

While the script is looking for Github releases

natifridman replied:
Looks like there is a [PR](https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/1998) to bump this dependency, but it's breaking some tests over there.
 I tried to use v1.4.1 in the go.mod, but I got errors that this [PR](https://github.com/kubernetes-sigs/gateway-api/pull/4357) in the gateway-api should fix

natifridman replied:
In the meantime I can modify the `manage.gateway-api-crd.sh` script to find the next release (1.4.0) if it can't find the release in Github.
@terrytangyuan WDYT?

terrytangyuan replied:
> In the meantime I can modify the manage.gateway-api-crd.sh script to find the next release (1.4.0) if it can't find the release in Github.

Sounds good

natifridman replied:
I added this change, now it should fallback to the next release

natifridman committed:
Fix CI tests failures

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Fix precommit CI failure

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Fix LLMISvc E2E test failure during rolling update

Use kubectl rollout status instead of kubectl wait --for=condition=available
to ensure the rolling update completes before running tests. The previous
approach would succeed as soon as any pod was ready, causing tests to run
while the old controller (without v1alpha2 webhook endpoints) was still
handling requests.

Also add v1alpha2 paths to workflow triggers to ensure E2E tests run when
v1alpha2 code is modified.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Update E2E tests for GIE v1.2.0 and fix flake8 lint error

- Update InferencePool backendRefs API group from
  inference.networking.x-k8s.io to inference.networking.k8s.io
  to match the graduated API in Gateway Inference Extension v1.2.0
- Remove unused global statement in kserve_storage.py to fix flake8 F824

Signed-off-by: Nati Fridman <nafridma@redhat.com>

sbekkerm committed:
Set kv-cache metric to vllm:kv_cache_usage_perc

Refs: RHOAIENG-41868

Explicitly set --kv-cache-usage-percentage-metric so EPP does not rely on
the legacy default (vllm:gpu_cache_usage_perc) from released GIE versions.

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
fix: Add v1alpha2 webhook support for LLMInferenceService

The install manifests were missing webhook configurations for
v1alpha2 API versions of LLMInferenceService and LLMInferenceServiceConfig.
Additionally, webhooks used matchPolicy: Equivalent which caused the API
server to convert v1alpha2 resources to v1alpha1 before validation,
resulting in schema mismatch errors.

Changes:
- Add v1alpha2 webhook entry for LLMInferenceService validation
- Add v1alpha2 webhook entry for LLMInferenceServiceConfig validation
- Update webhook names to include version suffix (.v1alpha1.validator,
  .v1alpha2.validator) for clarity
- Set matchPolicy: Exact on all webhooks to ensure each API version is
  validated with its correct schema types

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

sbekkerm committed:
Revert v1alpha2 validating webhooks from v0.16.0

Remove duplicate v1alpha2 ValidatingWebhookConfiguration entries for
LLMInferenceService and LLMInferenceServiceConfig that should be
auto-generated during the release process

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

natifridman committed:
Add fallback for Gateway API pseudo-versions in CRD install script

When gateway-api dependency uses a pseudo-version (e.g.,
v1.3.1-0.20251106052652-079e4774d76b), the CRD install script fails
because pseudo-versions don't have corresponding GitHub release tags.

This adds a resolve_gateway_api_version() function that:
- Detects pseudo-versions using regex pattern matching
- Fetches available releases from GitHub API
- Finds the next available release >= the base version
- Falls back gracefully if the API call fails

For example, v1.3.1-0.20251106052652-079e4774d76b resolves to v1.4.0.

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Sync Helm chart llm-d image versions with config directory

The Helm charts in charts/kserve-llmisvc-resources/templates/ were not
updated when the config/llmisvcconfig/ files were updated with new
llm-d image versions. This caused E2E test failures because the tests
use USE_LOCAL_CHARTS=true, which deploys from the Helm charts with
outdated v0.2.x images instead of the v0.4.0 images required for
GIE v1.2.0 compatibility.

Updated images:
- llm-d-dev:v0.2.2 → llm-d-cuda:v0.4.0
- llm-d-routing-sidecar:v0.2.0 → v0.4.0

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
bump  Envoy AI Gateway to v0.4.0

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
Update Envoy Gateway installation for AI Gateway v0.4.0 compatibility

The Envoy AI Gateway v0.4.0 changed its installation method from kubectl
apply of individual config files to Helm values files. This update:
- Downloads envoy-gateway-values.yaml and inference-pool addon from
  the AI Gateway repo
- Passes values files to Envoy Gateway helm install via -f flags
- Removes obsolete kubectl apply commands for redis.yaml, config.yaml,
  rbac.yaml, and inference-pool/config.yaml (paths no longer exist)

This is required for GIE v1.2.0 compatibility since AI Gateway v0.4.0
adds support for the graduated InferencePool API (inference.networking.k8s.io/v1)
instead of the experimental API (inference.networking.x-k8s.io/v1alpha2).

Signed-off-by: Nati Fridman <nafridma@redhat.com>

natifridman committed:
precommit changes

Signed-off-by: Nati Fridman <nafridma@redhat.com>

pierDipi commented on charts/kserve-llmisvc-resources/templates/webhookconfiguration.yaml:
```diff
@@ -12,7 +12,8 @@ webhooks:
     failurePolicy: Fail
-    name: llminferenceservice.kserve-webhook-server.validator
+    matchPolicy: Exact
+    name: llminferenceservice.kserve-webhook-server.v1alpha1.validator
```
by changing the name, an existing installation when upgrading will continue to have the webhook `llminferenceservice.kserve-webhook-server.validator` as a "ghost" (without match policy set), I think we need to try to keep the name for v1alpha1 validation(s) the same (for both llmisvc and llmisvc config down below)

sbekkerm replied:
Good catch - fixed in the last commit

sbekkerm committed:
fix: preserve original v1alpha1 webhook names to prevent ghost webhooks on upgrade

Reverts the v1alpha1 webhook names to their original format (without version
suffix) to ensure existing installations upgrade cleanly

Signed-off-by: Sergey Bekkerman <sbekkerm@redhat.com>

KillianGolds commented on pkg/apis/serving/v1alpha1/llm_inference_service_conversion.go:
Looks good overall. Nice clean standard Hub Spoke conversion pattern. My only minor concern is the silent error fallback on InferencePoolSpec conversion, but the comment acknowledges it shouldn't happen in practice so maybe just a nit.

KillianGolds commented on pkg/apis/serving/v1alpha1/llm_inference_service_conversion.go:
```diff
@@ -0,0 +1,471 @@
+	srcPool := &igwapiv1alpha2.InferencePool{Spec: *src}
+	dstPool := &igwapiv1.InferencePool{}
+
+	if err := srcPool.ConvertTo(dstPool); err != nil {
+		// Fallback: return empty spec on error (should not happen in practice)
+		return &igwapiv1.InferencePoolSpec{}
+	}
```
Could the silent fallback hide potential issues ? Would logging/propagating it be worth it?

KillianGolds commented on pkg/apis/serving/v1alpha1/llm_inference_service_conversion.go:
```diff
@@ -0,0 +1,471 @@
+	srcPool := &igwapiv1.InferencePool{Spec: *src}
+	dstPool := &igwapiv1alpha2.InferencePool{}
+
+	if err := dstPool.ConvertFrom(srcPool); err != nil {
+		// Fallback: return empty spec on error (should not happen in practice)
+		return &igwapiv1alpha2.InferencePoolSpec{}
+	}
```
Same as previous silent fallback comment.

KillianGolds commented on pkg/controller/v1alpha2/llmisvc/scheduler.go:
Looking good. The changes are straightforward, moving from GIE v1alpha2 to v1, removing the InferenceModel (which GIE v1 doesn't use in the same way), and updating RBAC. The dual API group in RBAC is good for backward compatibility.

KillianGolds replied:
### Follow-up: Zero-Downtime Migration
This PR does a clean break to GIE v1. For production environments, a zero-downtime migration approach (dual-pool creation with weighted HTTPRoute backendRefs) would be a valuable follow-up contribution.

KillianGolds commented on pkg/controller/v1alpha2/llmisvc/router_discovery.go:
```diff
@@ -350,10 +351,17 @@ func nonReadyHTTPRouteTopLevelCondition(route *gwapiv1.HTTPRoute) (*metav1.Condi
 // They must be accepted by their parent Gateways to be considered operational
 func IsInferencePoolReady(pool *igwapi.InferencePool) bool {
-	if pool == nil || len(pool.Status.Parents) == 0 {
+	if pool == nil {
 		return false
 	}
 
+	// If no parents have been set, consider the pool ready if it exists and has a valid spec
+	// This handles cases where no Gateway controller is populating the status
+	if len(pool.Status.Parents) == 0 {
+		// Pool is ready if it exists with a valid selector and target ports
+		return len(pool.Spec.Selector.MatchLabels) > 0 && len(pool.Spec.TargetPorts) > 0
+	}
```
This seems like a reasonable fix for the InferencePool readiness check. If I'm understanding correctly, the alternative would be blocking users whose gateway controller doesn't populate `Status.Parents`, which seems worse.

One thing I noticed, with this fallback, a pool could briefly be considered "ready" before the gateway has actually processed it (since we're only checking the spec exists, not that the routing is set up). Probably a short window and acceptable tradeoff, but wanted to flag it. Let me know if I'm misunderstanding the scenario.

KillianGolds commented on pkg/controller/v1alpha2/llmisvc/router.go:
Nice addition, validation failure shortcircuit in `EvaluateGatewayConditions` and `EvaluateHTTPRouteConditions` prevents overwriting validation errors.

KillianGolds reviewed (commented):
LGTM overall. The implementation is clean and aligns with similar work we did downstream. The conversion webhooks, RBAC changes, and InferencePool readiness fallback all look reasonable.

### Follow-up: Zero-Downtime Migration

This PR does a clean break to GIE v1 which is the right approach for simplicity. For production environments upgrading from existing v1alpha2 deployments, we might follow up with a zero-downtime migration strategy like I was working on downstream:

- Dual-pool creation (both v1 and v1alpha2 InferencePools)
- Readiness Check for Both Pools
- Weighted HTTPRoute backendRefs to gradually shift traffic
- One-way migration annotation to prevent flapping

Happy to contribute this as a follow-up PR

natifridman committed:
Precommit go.sum changes

Signed-off-by: Nati Fridman <nafridma@redhat.com>
